Support multiple input dims for sharded linear.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70266

Addresses some of the issues mentioned in
https://github.com/pytorch/pytorch/issues/65638. ShardedLinear implementation
only support 2D inputs.

On the other hand `nn.Linear` supports arbitrary dimensions for inputs and
outputs. As a result, in this PR I've added support to ensure that
ShardedLinear supports arbitrary input dims as well.

Differential Revision: [D33267630](https://our.internmc.facebook.com/intern/diff/D33267630/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

pritamdamania committed:
Support multiple input dims for sharded linear.

Addresses some of the issues mentioned in
https://github.com/pytorch/pytorch/issues/65638. ShardedLinear implementation
only support 2D inputs.

On the other hand `nn.Linear` supports arbitrary dimensions for inputs and
outputs. As a result, in this PR I've added support to ensure that
ShardedLinear supports arbitrary input dims as well.

Differential Revision: [D33267630](https://our.internmc.facebook.com/intern/diff/D33267630/)

[ghstack-poisoned]

fduwjj reviewed (commented):
N00b question. Under what scenario, do people need to use more than 2 dim in the input?

Overall, this looks great. My questions is mainly around the way of implementation not correctness. If this is urgent. I can stamp it.

wanchaol reviewed (approved):
lgtm

pritamdamania87 commented on torch/distributed/_sharded_tensor/ops/linear.py:
```diff
@@ -177,7 +177,9 @@ def _handle_row_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
-    input_t = input.t().contiguous()
+    input_size = list(reversed(range(input.dim())))
+
+    input_t = input.permute(input_size).contiguous()
```
Good point, fixed this.

pritamdamania committed:
Update on "Support multiple input dims for sharded linear."


Addresses some of the issues mentioned in
https://github.com/pytorch/pytorch/issues/65638. ShardedLinear implementation
only support 2D inputs.

On the other hand `nn.Linear` supports arbitrary dimensions for inputs and
outputs. As a result, in this PR I've added support to ensure that
ShardedLinear supports arbitrary input dims as well.

Differential Revision: [D33267630](https://our.internmc.facebook.com/intern/diff/D33267630/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

pritamdamania87 commented:
> N00b question. Under what scenario, do people need to use more than 2 dim in the input?

Really depends on how the user represents their features. For example, it could be an RGB image as a 4D tensor (1 additional dimension for batch) that goes through multiple conv layers and then fed into a linear layer.
